relay-server/src/actors/spooler/mod.rs

```diff
@@ -525,33 +730,35 @@ impl BufferService {
             sender,
         } = message;
 
-        for key in &keys {
-            for value in self.buffer.remove(key).unwrap_or_default() {
-                self.used_memory -= value.estimated_size();
-                sender.send(value).ok();
+        match self.state {
+            BufferState::Memory { ref mut ram }
+            | BufferState::MemoryFileStandby { ref mut ram, .. } => {
+                ram.dequeue(&keys, sender);
+            }
+            BufferState::MemoryFileRead {
+                ref mut ram,
+                ref disk,
+            } => {
+                ram.dequeue(&keys, sender.clone());
+                disk.dequeue(project_key, &mut keys, sender, &self.services)
+                    .await;
+
+                // If there is nothing left on the disk, just swithc to in-memory buffer.
+                if disk.is_empty().await.unwrap_or_default() {
+                    self.state =
+                        BufferState::to_memory(self.config.spool_envelopes_max_memory_size(), disk);
+                }
+            }
+            BufferState::File { ref disk } => {
+                disk.dequeue(project_key, &mut keys, sender, &self.services)
+                    .await;
+
+                // If there is nothing left on the disk, just swithc to in-memory buffer.
+                if disk.is_empty().await.unwrap_or_default() {
+                    self.state =
+                        BufferState::to_memory(self.config.spool_envelopes_max_memory_size(), disk);
```

Comment:
This and similar logic (check some condition -> transition state) seems to be duplicated in a few places. Can we instead have an `update_state(&mut self)` method where all transitions are handled? Beside reducing code duplication, this method would then compactly represent the entire state machine, which makes understanding it more easy.